### DISCUSSION TITLE
DRY code

### DISCUSSION THREAD

**Discussion by pedrodsrodrigues** [apple/pkl#293] [Q&A] [answered]
Take this example in Terraform, for example:
```
resource "kubectl_manifest" "application_deployment" {
  yaml_body = templatefile("${path.module}/application/deployment.yaml")
}

resource "kubectl_manifest" "application_two_deployment" {
  yaml_body = templatefile("${path.module}/application_two/deployment.yaml")
}
```

How could I make it dynamic, using PKL? Or something like this isn't part of what PKL is trying to solve?

Some answers would be very appreciated since I currently have a lot of repetition in my code.
Thank you!

**Comment by holzensp** [comment]
It's hard to know what you're optimising for. In this case, you're importing two different files. That, by itself, is not going to be much more compact, unless you're just looking for `import*("./*/deployment.pkl")`. Can you be a little more specific with your question?

**Reply by pedrodsrodrigues** [reply]
I'm sorry if I wasn't clear enough, and thank you for the answer. I hope this makes the objective more clear:
I have 2 applications, and the only difference between both deployments is the name (they have the same structure, just different values). The code would be reduced since I would simply say that I have `application` and `application_two`, and the Terraform resource would use a variable instead of having the text hardcoded.
As if I simply had `List("application", "application_two")`, and then I would loop through those values to create a Terraform resource. Almost as if the Terraform resource would be something like the following:
```
resource "kubectl_manifest" "${projects}_deployment" {
  yaml_body = templatefile("${path.module}/${projects}/deployment.yaml")
}
```

I don't want to stop using Terraform. I want PKL to give instructions to Terraform.

**Reply by holzensp** [reply, marked as answer]
To use Pkl with Terraform, I recommend using their `.tf.json` format. Still, your example is hard to DRY-up, considering it just consumes one file per app; the only thing that will DRY that up is using glob imports.

**Comment by Apollo-XIV** [comment]
I know you got an answer to this a while ago, but have you considered using a for_each block to do this completely in terraform?
```hcl
resource "kubectl_manifest" "deployments" {
    for_each = [
        "application_one",
        "application_two"
    ]
    yaml_body = templatefile("${path.module}/${each.key}/deployment.yaml")
}
```
